drivers/espi/espi_emul.c

```diff
@@ -17,174 +17,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
 {
-	struct espi_emul_data *data = dev->data;
-	sys_snode_t *node;
+	unsigned int i;
+
+	for (i = 0; i < NUMBER_OF_VWIRES; i++) {
+		data->vw_state[i] = vw_state_default[i];
+	}
+}
 
-	SYS_SLIST_FOR_EACH_NODE(&data->emuls, node) {
-		struct espi_emul *emul;
+/**
+ * Find a virtual wire in the array placed in the host data.
+ *
+ * @param data Host emulator data with the vwire array
+ * @param vw Virtual wire signal to be found
+ * @return index in the array
+ * @return -1 if not found
+ */
+static int espi_emul_host_find_index(struct espi_emul_data *data, enum espi_vwire_signal vw)
+{
+	int idx;
 
-		emul = CONTAINER_OF(node, struct espi_emul, node);
-		if (emul->chipsel == chipsel) {
-			return emul;
+	for (idx = 0; idx < NUMBER_OF_VWIRES; idx++) {
+		if (data->vw_state[idx].sig == vw) {
+			return idx;
 		}
 	}
 
-	return NULL;
+	return -1;
 }
 
-static int espi_emul_config(const struct device *dev, struct espi_cfg *cfg)
+#ifdef CONFIG_ESPI_EMUL_KBC8042
+static void espi_emul_send_ibf_irq(const struct device *dev)
 {
 	struct espi_emul_data *data = dev->data;
+	struct espi_evt_data_kbc *kbc_evt;
+	struct espi_event ev;
 
-	__ASSERT_NO_MSG(cfg);
+	if (data->kbc8042_irq_enabled == false ||
+	    (data->kbc8042_status & KBC8042_STATUS_IBF) == 0) {
+		return;
+	}
 
-	data->cfg = *cfg;
+	ev.evt_type = ESPI_BUS_PERIPHERAL_NOTIFICATION;
+	ev.evt_details = ESPI_PERIPHERAL_8042_KBC;
 
-	return 0;
+	kbc_evt = (struct espi_evt_data_kbc *)&ev.evt_data;
+	kbc_evt->evt = HOST_KBC_EVT_IBF;
+	kbc_evt->data = data->kbc8042_dbbin;
+	kbc_evt->type = (data->kbc8042_status & KBC8042_STATUS_A2);
+
+	espi_send_callbacks(&data->callbacks, dev, ev);
+
+	/* Normal MCU clears the flag and register after reading from it.
+	 * We should follow this behavior.
+	 */
+	data->kbc8042_dbbin = 0;
+	data->kbc8042_status &= ~(KBC8042_STATUS_A2 | KBC8042_STATUS_IBF);
 }
 
-static int emul_espi_trigger_event(const struct device *dev, struct espi_event *evt)
+static void espi_emul_send_obe_irq(const struct device *dev)
 {
 	struct espi_emul_data *data = dev->data;
+	struct espi_evt_data_kbc *kbc_evt;
+	struct espi_event ev;
 
-	if (((evt->evt_type & ESPI_BUS_EVENT_VWIRE_RECEIVED) &&
-	     !(data->cfg.channel_caps & ESPI_CHANNEL_VWIRE)) ||
-	    ((evt->evt_type & ESPI_BUS_EVENT_OOB_RECEIVED) &&
-	     !(data->cfg.channel_caps & ESPI_CHANNEL_OOB)) ||
-	    ((evt->evt_type & ESPI_BUS_PERIPHERAL_NOTIFICATION) &&
-	     !(data->cfg.channel_caps & ESPI_CHANNEL_PERIPHERAL))) {
-		return -EIO;
+	if (data->kbc8042_irq_enabled == false || data->kbc8042_obe_irq_pending == false) {
+		return;
 	}
 
-	espi_send_callbacks(&data->callbacks, dev, *evt);
+	data->kbc8042_obe_irq_pending = false;
+
+	ev.evt_type = ESPI_BUS_PERIPHERAL_NOTIFICATION;
+	ev.evt_details = ESPI_PERIPHERAL_8042_KBC;
+
+	kbc_evt = (struct espi_evt_data_kbc *)&ev.evt_data;
+	kbc_evt->evt = HOST_KBC_EVT_OBE;
+	kbc_evt->data = 0;
+	kbc_evt->type = 0;
+
+	espi_send_callbacks(&data->callbacks, dev, ev);
+}
+#endif
```

Comment:
nit: this is a long `ifdef`, please add a closing `/* CONFIG_ESPI_EMUL_KBC8042 */